[refactor] 테마 컬러 리팩토링

nakisara01 opened the pull request:
### 설명
전체 View의 Color를 Palette화 해서 이후 테마 변경이 용이하도록 수정했습니다.

### 관련 이슈

Closes #168

### 작업 내용

- [ ] HomePalette 개발 및 HomeView 변경
- [ ] ProgressPalette 개발 및 ChordProgressView, Segment, Waveform, ChordFingeringView, GuitarFingeringView, PianoFingeringView 변경
- [ ] ExportPalette 개발 및 ExportView, ChordSheetView 변경
- [ ] RecordingPalette 개발 및 RecordingView, LoadingView 변경
- [ ] ComponentPalette 개발 및 ErrorView, BackButton, ControllerButton 변경

### 참고 내용

nakisara01 committed:
[#168] Fix: modify HomeView Color

HomePalette를 생성하고 이에 맞춰 HomeView Color 설정 방식을 변경했습니다.

nakisara01 committed:
[#168] Fix: modify RecordingView Style

RecordingPalette를 추가하고 RecordingView와 LoadingView를 이에 맞게 수정했습니다

nakisara01 committed:
[#168] Fix: modify ExportView style

ExportPalette를 생성하고 ExportView를 이에 맞게 적용했습니다

nakisara01 committed:
[#168] Feat: add ProgressPalette

nakisara01 committed:
[#168] Fix: modify ChordProgressView style

ChordProgressView를 ProgressPalette를 기반으로 색을 수정했습니다

nakisara01 committed:
[#168] Fix: add PianoFingering

PianoKey를 ProgressPalette에 추가하고 PianoFingeringView에 적용했습니다

nakisara01 committed:
[#168] Fix: add Component Palette

Component의 색을 ComponentPalette로 빼고 BackButton, ControllerButton, ErrorView에 적용했습니다

Segyun reviewed (commented):
저는 이런 식으로 선택된 팔레트에 따라서 색상을 바꿀 수 있는 것으로 생각했었어요!

```Swift
enum Palette {
  case blue
  case yellow
  ...
}

extension Palette {
  var button: Color {
    switch self {
      case .blue: .blue1
      case .yellow: .yellow
    }
  }
}

struct ChordProgressView: View {
  let palette: Palette
  
  var body: some View {
    Button("Some Button") { ... }
    .background(palette.button)
  }
}
```

ehdrbdndns reviewed (commented):
페이지 별 팔레트로 할 시 관리 차원에서 기존과 크게 이점이 없는 것 같긴 하네요!

nakisara01 committed:
[#168] refactor: migration palette

페이지 별로 분리된 팔레트를 palette에 통합하고, 테마 변경이 용이하도록 했습니다.
추가적으로 PaletteEnvironment와 ThemeManager를 추가해 테마를 앱 전체에서 주입받기 쉽도록 조정했습니다.

nakisara01 committed:
[#168] fix: apply palette on HomeView

nakisara01 committed:
[#168] fix: apply ExportView & ChordSheetView

nakisara01 committed:
[#168] fix: apply palette on Components

nakisara01 committed:
[#168] fix: apply palette ProgressView Section

nakisara01 committed:
[#168] fix: apply palette Recording & LoadingView

Segyun commented on GMG/Core/DesignSystem/Color/ThemeManager.swift:
```diff
@@ -0,0 +1,36 @@
+import Combine
+import SwiftUI
+
+final class ThemeManager: ObservableObject {
```
Envrionment에도 `@Observable` 사용할 수 있어요!

[Migrating from the Observable Object protocol to the Observable macro](https://developer.apple.com/documentation/SwiftUI/Migrating-from-the-observable-object-protocol-to-the-observable-macro)

```suggestion
@Observable
final class ThemeManager {
```

`GMGApp.swift`

```Swift
@State private var themeManager: ThemeManager = .init()

...

MainView()
    .environment(themeManager)
```

다른 뷰에서 사용 시

```Swift
@Environment(ThemeManager.self) private var themeManager
```

Segyun commented on GMG/Core/DesignSystem/Color/PaletteEnvironment.swift:
```diff
@@ -0,0 +1,14 @@
+
+import SwiftUI
+
+private struct PaletteKey: EnvironmentKey {
+    static let defaultValue: Palette = .basic  // 기본값은 basic
+}
+
+extension EnvironmentValues {
+    var palette: Palette {
+        get { self[PaletteKey.self] }
+        set { self[PaletteKey.self] = newValue }
+    }
+}
```
`@Entry` 매크로를 사용하면 더 짧게 표현할 수 있어요!

[Entry()](https://developer.apple.com/documentation/swiftui/entry())

```suggestion
extension EnvrionmentValues {
    @Entry var palette: Palette = .basic
}
```

Segyun commented on GMG/Core/DesignSystem/Color/Palette.swift:
제가 전에 제안 드린 방식처럼 팔레트 자체를 enum으로 정의하면 새로운 case를 추가하게 되면 각각의 프로퍼티의 switch 문에 새로운 case를 추가해야 되어서 번거로울 것 같긴 합니다. 그래서 다시 생각해보았을 때 팔레트를 구조체로 만들고, 정적 변수로 `basic`, `next` 등의 케이스를 추가하는 것이 어떨까 싶습니다! 새로운 팔레트 케이스를 추가할 때, 이전 팔레트는 건들지 않으면서도 확장할 수 있어서 더 좋을 것 같습니다.

```Swift
struct Palette {
    let backgroundColor: Color
    let sheetBackground: Color
    ...
}

extension Palette {
    static let basic: Palette = .init(
        backgroundColor: .bg1,
        sheetBackground: .white,
        ...
    )
    
    static let next: Palette = .init(
        backgroundColor: .bg1,
        sheetBackground: .white,
        ...
    )
}
```

또한 만약 뷰별로 팔레트를 구분해야 된다면 다음과 같은 식으로 분리하여 구현이 가능할 것 같습니다!

```Swift
struct Palette {
    let homePalette: HomePalette
    let exportPalette: ExportPalette
}

extension Palette {
    static let basic: Palette = .init(
        homePalette: .basic,
        exportPalette: .basic,
    )
}

struct HomePalette {
    let primaryButtonText: Color
}

extension HomePalette {
    static let basic: HomePalette = .init(
        primaryButtonText: .black
    )
}
```

nakisara01 replied:
해당 작업을 새로 PR을 파서 진행해보겠습니다!

Segyun reviewed (approved):
고생하셨습니다~

가능하면 모든 컴포넌트별로 색상을 따로 지정하는 것이 세세하게 커스텀할 수 있어서 좋을 것 같긴 한데 너무 시간이 많이 드는 작업일 것 같아서 일단은 같은 색상끼리는 묶어서 정의하고 특수하게 변경이 필요하게 된다면 그때 추가하는 것이 어떨까 싶습니다!

그리고 만약 테마에 따라 버튼 색상만 바뀐다면 팔레트에는 버튼 색상만 정의해도 괜찮을 것 같아요!

ehdrbdndns commented on GMG/Core/DesignSystem/Color/Palette.swift:
고생하셨습니다!

막상 코드를 보니 이런 방식이 오히려 코드가 더 복잡해지지는 않나 우려됩니다.

팔레트는 결국 디자인에 맞춰서 짜야하는데, 디자이너분들이 디자인해주실 때 컴포넌트화 없이 막 하셨기에 디자인에 대한 네이밍이 전혀 없는 상황입니다.

지금 네이밍 된 것은 결국 조이드가 혼자 열심히 구상한 네이밍이지만, 사실 레몬과 저는 이 네이밍을 보고 바로 어디 컴포넌트에 쓰인 style인지 파악하기 어렵습니다. 팔레트가 제대로 쓰일려면 우선 피그마에 있는 컴포넌트들을 컴포넌트화 시키고 각자 네이밍을 조이드가 만든 네이밍에 맞게끔 적용시켜야 이 팔레트가 의미가 있다고 생각합니다. 

그렇지 않을 경우 추후에 새로운 컴포넌트를 추가한다고 했을 때, 팔레트에 있는 코드를 재활용할 수 있는 것은 조이드만 제대로 활용할테고 레몬과 저는 오히려 시간만 더 들 것 같다는 생각이네요.

----

포폴용으로 사용하신다면 네이밍을 피그마 컴포넌트의 네이밍과 맞췄다고 이야기할 필요성이 있어보입니다.

nakisara01 committed:
Merge branch 'dev' into refactor/#168

nakisara01 committed:
[#168] fix: change mlpackage
